File: rvs/m-files/plotCDFci.m
function [x_avg, x_med, x_sd, x_cov] = plotCDFci(x,ci,figNo)
% [x_avg, x_med, x_sd, x_cov] = plotCDFci(x,ci,figNo)
%
% plot a CDF with confidence intervals
% 
%    INPUT         DESCRIPTION                                DIMENSION
%    =====         ================================           ================
%     x            a vector of a random sample                n x 1  or 1 x n
%     ci           a confidence level, eg 90 or 95            1 x 1
%    figNo         figure number                              1 x 1 
%   OUTPUT         DESCRIPTION                                DIMENSION
%   -=====         ================================           ================
%    x_avg         the average value of x                     1 x 1
%    x_med         the median  value of x                     1 x 1
%    x_sd          the standard deviation of x                1 x 1
%    x_cov         the coefficient of variation of x          1 x 1

 pdfPlots = 0;

 N = length(x);                         % number of values in the sample
 x = sort(x);                           % sort the sample
 x_avg = sum(x)/N;
 x_med = x(round(N/2));
 x_sd  = sqrt(sum((x-x_avg).^2)/(N-1));
 x_cov = abs(x_sd/x_avg);
 nBins = floor(N/5);                    % number of bins in the histogram
 [fx,xx] = hist(x,nBins);               % compute the histogram
  fx = fx / N * nBins/(max(x)-min(x));  % scale the histogram to a PDF
  Fx = ([1:N])/(N+1);                   % empirical CDF
 sFx = sqrt(Fx.*(1-Fx)/(N+2));          % standard error of Fx Gumbel p.47
%if x(N) > 1, 
%  probability_X_gt_1 = sum(x>1) / (N+1) % fraction of the sample for which X>1
%end


if figNo 

 ppl = [ 68, 1, 84 ]/256;    % purple to match histogram bars
 patchColor = [ 0.80 , 0.85 , 1 ];
 figure(figNo)
  clf
  z   = -norm_inv( (1-ci/100)/2 );
  Fps = min(Fx + z*sFx,1-0.001/N);
  Fms = max(Fx - z*sFx,0+0.001/N);
  [ xsp , Fsp ] = stairs( (x), Fps);
  [ xsm , Fsm ] = stairs( (x), Fms);
  xp = [ xsp ; xsm(end:-1:1) ; xsp(1) ];
  Fp = [ Fsp ; Fsm(end:-1:1) ; Fsp(1) ];
  hold on
  hc = patch(xp, Fp,  patchColor);
  hm = plot( x_med*[1,1,0] , 0.5*[0,1,1] , '--r', 'LineWidth', 2 );
  set(hc, 'EdgeColor', patchColor );
  ee = 0.05;
  xl =  (1+ee)*x(1) - ee*x(N);
  xh = -ee*x(1)     + (1+ee)*x(N);
  hx =  0.30*xl     + 0.7*xh;
% if  x(1) < 1 && x(N) > 1 
%   plot( [1,1,0] , (1-probability_X_gt_1)*[0,1,1] , '--k', 'LineWidth', 2 );
%   text (hx,0.5, sprintf('P[X>1] = %5.3f', probability_X_gt_1 )) 
% end
  text( x_med-0.05,-0.07, 'x_{med}', 'color','r')
  text (hx,0.4, sprintf('x_{avg} = %5.3f',x_avg))
  text (hx,0.3, sprintf('x_{med} = %5.3f',x_med))
  text (hx,0.2, sprintf('x_{sd}   = %5.3f',x_sd))
  if x_cov < 100
    text (hx,0.1, sprintf('x_{cov}  = %5.3f',x_cov))
  end
% title(sprintf(' x1 =  %f ', x(1)))

  stairs( x, Fx, 'color', ppl, 'LineWidth', 3.0 ); % plot empirical CDF + c.i.
  axis( [ xl  xh 0 1.05 ]);
  xlabel('x')
  ylabel('CDF       F_X(x)  =  Prob[ X<=x ]')

 if pdfPlots, print('plotCDFci.pdf','-dpdfcrop'); end

end

File: multivarious/utl/plot_CDF_ci.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.special import betaincinv
from scipy.stats import norm

def plot_CDF_ci(data, confidence_level, fig_num, x_label='sorted sample values', norm_inv_CDF=False ):
    """
    Plot empirical CDF of data with confidence intervals.
    
    Parameters
    ----------
    data : array_like
        Data values
    confidence_level : float
        Confidence level (e.g., 95 for 95%)
    fig_num : int
        Figure number
    x_label : string
        label for the x axis
    norm_inv_CDF : boolean
        True: plot norm inverse of F , False: plot F (default)
    """

    #plt.rcParams['text.usetex'] = True # Set to True if LaTeX is installed

    pdf_plots = True  # Set to True to save PDF files
    interactive = True # Enable interactive mode for matplotlib
    
    if interactive:
        plt.ion() # plot interactive mode: on

    data = np.asarray(data).flatten()
    N = len(data)
    
    # Empirical CDF
    Fx = np.arange(1, N + 1) / ( N + 1 )    # Gumbel (1958)

    # Sort data
    sorted_data = np.sort(data)
    
    # Confidence intervals on Fx
    # The variance of the emperical CDF is (F)(1-F)/(N+2)   # Gumbel (1958)
    alpha = 1 - confidence_level / 100
    P = 1 - alpha/2 

    # --- simultaneous, conservative: using Dvoretzky-Kiefer-Wolfowitz inequality
    # https://en.wikipedia.org/wiki/Dvoretzky-Kiefer-Wolfowitz_inequality
    '''
    epsilon = np.sqrt(np.log(2 / alpha) / (2 * N))
    upper_ci_DKW = np.minimum(Fx + epsilon, 1.0 - 0.001 / N)
    lower_ci_DKW = np.maximum(Fx - epsilon, 0.0 + 0.001 / N)
    '''
    
    # --- simultaneous, exact: Dirichlet-based simultaneous emperical CDF bands
    '''
    M = 30000
    G = np.random.gamma(1.0, 1.0, size=(M, N+1))
    D = G / G.sum(axis=1, keepdims=True)
    U = D.cumsum(axis=1)[:, :N]
    lower_ci_D = np.quantile(U, alpha/2, axis=0)
    upper_ci_D = np.quantile(U, 1-alpha/2, axis=0)
    '''

    # --- pointwise, approximate: using Gumble (p.47) and normal distribution
    '''
    sFx = np.sqrt(Fx*(1-Fx)/(N+2))             # standard error of Fx Gumbel p.47
    z   = norm.ppf(P) 
    epsilon = z*sFx
    upper_ci_A = np.minimum(Fx + epsilon, 1.0 - 0.001 / N)
    lower_ci_A = np.maximum(Fx - epsilon, 0.0 + 0.001 / N)
    '''
    
    # --- pointwise, exact: Beta-based empirical CDF bands
    q = np.arange(1,N+1)
    p = N+1-q
    upper_ci_B = betaincinv(q, p,   P)  # upper quantile, near zero ( ci > 0 )
    lower_ci_B = betaincinv(q, p, 1-P)  # lower quantile, near one  ( ci < 1 )

    # a normal CDF for testing 
    #mu, std = np.mean(data), np.std(data)
    #x_theory = np.linspace(min(sorted_data), max(sorted_data), 200)
    #z_theory = (x_theory - mu) / std
    #cdf_theory = norm.cdf(z_theory)  
    
    # Plotting ---------------------------------------------------
    fig = plt.figure(fig_num, figsize=(10, 7))
    fig.clf()

    # select which confidence interval to plot
    lower_ci = lower_ci_B
    upper_ci = upper_ci_B
    
    if norm_inv_CDF:

       # apply the standard normal inverse function to Fx and its CI's 
       lower_ci = norm.ppf(lower_ci)   
       upper_ci = norm.ppf(upper_ci)
       Fx       = norm.ppf(Fx)
 
       maxCI = max(upper_ci)
       minCI = min(lower_ci)
       norm_F_val = np.arange( np.floor(minCI) , np.ceil(maxCI)+1 )
       norm_F_val = norm_F_val[ (-4 < norm_F_val) & (norm_F_val < 4) ]
       norm_F_str = np.char.mod('%6.4f',  norm.cdf(norm_F_val))
 
       #print(f' minCI = {minCI} ')  # debug
       #print(f' maxCI = {maxCI} ')  # debug
       #print(f' norm_F_val = {norm_F_val}')
       #print(f' norm_F_str = {norm_F_str}')

    plt.fill_between(sorted_data, lower_ci, upper_ci, 
                     color=[0.8, 0.9, 1.0], alpha=0.5,
                     label=f'{confidence_level}% confidence band')
    plt.step(sorted_data, Fx, '-', color=[0.2, 0.4, 0.8], 
             linewidth=2, markersize=4, label='Empirical CDF')
    # plot other CI's 
    '''
    plt.plot(sorted_data, upper_ci, '-r')
    plt.plot(sorted_data, lower_ci, '-r')
    plt.step(sorted_data, upper_ci_D, '-r' ) 
    plt.step(sorted_data, lower_ci_D, '-r' ) 
    plt.plot(sorted_data, upper_ci_DKW, '-g')
    plt.plot(sorted_data, lower_ci_DKW, '-g')
    '''
    # plot another CDF
    '''
    plt.plot(x_theory, cdf_theory, 'r-', linewidth=2, 
             label=f'Normal CDF (μ={mu:.3f}, σ={std:.3f})')
    plt.plot(sorted_data, sorted_data * 0 + 0.5, '--k', alpha=0.3, linewidth=1)
    '''
    
    plt.xlabel(x_label, fontsize=16)
    plt.ylabel('Cumulative probability', fontsize=16)
    if norm_inv_CDF == True:
        if maxCI > 3.5 or minCI < -3.5:
            plt.ylabel(r'$\Phi^{-1}$( Cumulative probability )', fontsize=16)
        else: 
            plt.ylabel('Cumulative probability', fontsize=16)
            plt.yticks(norm_F_val, norm_F_str)
    plt.title(f'Empirical CDF with {confidence_level}% Confidence Intervals', 
              fontsize=16)
    plt.legend(fontsize=16)
    plt.legend(loc='lower right')
    plt.grid(True, alpha=0.3)
    if norm_inv_CDF == False:
        plt.ylim([0, 1])
    
    plt.tight_layout()

    # Display plots
    if not interactive:
        plt.show()

    # Save plots to .pdf
    if pdf_plots:
        filename = f'plot_CDF_ci-{fig_num:04d}.pdf'
        plt.savefig(filename, bbox_inches='tight', dpi=300)
        print(f"Saved: {filename}")

    if interactive: 
        input("Press Enter to close all figures...")
        plt.close('all')

    '''
    Dvoretzky, A., Kiefer, J., Wolfowitz, J. (1956),
    "Asymptotic minimax character of the sample distribution function and of
    the classical multinomial estimator",
    Annals of Mathematical Statistics, 27 (3): 642-669.
    doi:10.1214/aoms/1177728174, MR 0083864

    Gumbel, Emil Julius (1958)
    Statistics of Extremes. Columbia University Press, New York. 

    Shore, H. (1982).
    Approximations to the inverse cumulative normal function for use
    on hand calculator. Applied Statistics, 28, 175-176.

    Tocher K. D. (1963).
    The Art of Simulation. English Universities Press, London.
    '''
